pokemongo_bot/walkers/step_walker.py

```diff
@@ -1,154 +1,82 @@
-          - `start_lat in decimal degrees
-          - `start_lng in decimal degrees
-          - `dest_lat in decimal degrees
-          - `dest_lng in decimal degrees
-    
-        :Returns:
-          The bearing in degrees
-    
-        :Returns Type:
-          float
-        """
-    
-        lat1 = math.radians(start_lat)
-        lat2 = math.radians(dest_lat)
-    
-        diffLong = math.radians(dest_lng - start_lng)
-    
-        x = math.sin(diffLong) * math.cos(lat2)
-        y = math.cos(lat1) * math.sin(lat2) - (math.sin(lat1)
-                * math.cos(lat2) * math.cos(diffLong))
-    
-        initial_bearing = math.atan2(x, y)
-    
-        # Now we have the initial bearing but math.atan2 return values
-        # from -180° to + 180° which is not what we want for a compass bearing
-        # The solution is to normalize the initial bearing as shown below
-        initial_bearing = math.degrees(initial_bearing)
-        compass_bearing = (initial_bearing + 360) % 360
-    
-        return compass_bearing
-        
-    def _get_next_pos(self, lat, lon, bearing, speed, precision):
-        origin = Point(lat, lon)
-        if speed == 0.0:
-            return lat, lon
-        if speed == float("inf"):
-            return self.destLat, self.destLng
+        return self.is_arrived()
+
+    def is_arrived(self):
+        inverse = Geodesic.WGS84.Inverse(self.bot.position[0], self.bot.position[1], self.dest_lat, self.dest_lng)
+        return inverse["s12"] <= self.precision + self.epsilon
+
+    def get_next_position(self, origin_lat, origin_lng, origin_alt, dest_lat, dest_lng, dest_alt, distance):
+        inverse = Geodesic.WGS84.Inverse(origin_lat, origin_lng, dest_lat, dest_lng)
+        total_distance = inverse["s12"]
+
+        if total_distance == 0:
+            total_distance = self.precision or self.epsilon
+
+        if distance == 0:
+            if not self.saved_location:
+                self.saved_location = origin_lat, origin_lng, origin_alt
+
+            dest_lat, dest_lng, dest_alt = self.saved_location
+            travel = self.precision
         else:
-            offset_angle = (1/self.speed)*(precision/1.74)
-            lat, lon, _ = VincentyDistance(kilometers=speed*1e-3).destination(origin, bearing + uniform(-offset_angle, offset_angle))
-            return lat, lon
+            self.saved_location = None
+            travel = min(total_distance, distance)
+
+        direct = Geodesic.WGS84.Direct(origin_lat, origin_lng, inverse["azi1"], travel)
+        next_lat = direct["lat2"]
+        next_lng = direct["lon2"]
+
+        random_azi = uniform(inverse["azi1"] - 90, inverse["azi1"] + 90)
+        random_dist = uniform(0.0, self.precision)
+        direct = Geodesic.WGS84.Direct(next_lat, next_lng, random_azi, random_dist)
+
+        next_lat = direct["lat2"]
+        next_lng = direct["lon2"]
+        next_alt = origin_alt + (travel / total_distance) * (dest_alt - origin_alt) + random_alt_delta()
```

Comment:
A point in between is not good because.. lets say we have below polyline

O...a...b...c...d..e.....z....D
now elevations are:
O- 404m
a - 490m
b - 420m
...
w - 530m
z - 440m
D - 404m